// novu/dist/constants/constants.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.getServerPort = exports.setAvailablePort = exports.SEGMENTS_WRITE_KEY = exports.ANALYTICS_ENABLED = exports.GITHUB_DOCKER_URL = exports.CLIENT_LOGIN_URL = exports.API_TRIGGER_URL = exports.API_NOTIFICATION_TEMPLATES_URL = exports.API_NOTIFICATION_GROUP_URL = exports.API_SWITCH_ENVIRONMENT_FORMAT_URL = exports.API_ENVIRONMENT_KEYS = exports.API_ENVIRONMENT_ME_URL = exports.API_CREATE_ENVIRONMENT_URL = exports.API_SWITCH_ORGANIZATION_FORMAT_URL = exports.API_CREATE_ORGANIZATION_URL = exports.API_UPDATE_EMAIL = exports.API_AUTH_SIGNUP = exports.API_OAUTH_URL = exports.EMBED_PATH = exports.TRIGGER_ROUTE = exports.WIDGET_DEMO_ROUTE = exports.REDIRECT_ROUTE = exports.SERVER_HOST = void 0;
const getPort = require("get-port");
const dotenv = require("dotenv");
dotenv.config();
exports.SERVER_HOST = 'localhost';
exports.REDIRECT_ROUTE = '/cli/callback';
exports.WIDGET_DEMO_ROUTE = '/demo';
exports.TRIGGER_ROUTE = '/trigger';
exports.EMBED_PATH = (_a = process.env.NOVU_EMBED_PATH) !== null && _a !== void 0 ? _a : 'https://embed.novu.co/embed.umd.min.js';
const apiAddress = (_b = process.env.NOVU_API_ADDRESS) !== null && _b !== void 0 ? _b : 'https://api.novu.co';
exports.API_OAUTH_URL = `${apiAddress}/v1/auth/github`;
exports.API_AUTH_SIGNUP = `${apiAddress}/v1/auth/register`;
exports.API_UPDATE_EMAIL = `${apiAddress}/v1/users/profile/email`;
exports.API_CREATE_ORGANIZATION_URL = `${apiAddress}/v1/organizations`;
exports.API_SWITCH_ORGANIZATION_FORMAT_URL = `${apiAddress}/v1/auth/organizations/{organizationId}/switch`;
exports.API_CREATE_ENVIRONMENT_URL = `${apiAddress}/v1/environments`;
exports.API_ENVIRONMENT_ME_URL = `${apiAddress}/v1/environments/me`;
exports.API_ENVIRONMENT_KEYS = `${apiAddress}/v1/environments/api-keys`;
exports.API_SWITCH_ENVIRONMENT_FORMAT_URL = `${apiAddress}/v1/auth/environments/{environmentId}/switch`;
exports.API_NOTIFICATION_GROUP_URL = `${apiAddress}/v1/notification-groups`;
exports.API_NOTIFICATION_TEMPLATES_URL = `${apiAddress}/v1/notification-templates`;
exports.API_TRIGGER_URL = `${apiAddress}/v1/events/trigger`;
exports.CLIENT_LOGIN_URL = process.env.NOVU_CLIENT_LOGIN || 'https://web.novu.co/auth/login';
exports.GITHUB_DOCKER_URL = 'https://github.com/novuhq/novu/tree/main/docker';
exports.ANALYTICS_ENABLED = process.env.ANALYTICS_ENABLED === 'false' ? false : true;
exports.SEGMENTS_WRITE_KEY = process.env.CLI_SEGMENT_WRITE_KEY || 'tz68K6ytWx6AUqDl30XAwiIoUfr7iWVW';
let SERVER_PORT;
async function setAvailablePort() {
    SERVER_PORT = await getPort();
}
exports.setAvailablePort = setAvailablePort;
async function getServerPort() {
    if (!SERVER_PORT) {
        throw new Error('setAvailablePort must be called a single time to generate available port');
    }
    return SERVER_PORT;
}
exports.getServerPort = getServerPort;
